digiroad2-oracle/src/main/scala/fi/liikennevirasto/digiroad2/TierekisteriClient.scala

```diff
@@ -157,24 +157,187 @@ case class TierekisteriMassTransitStop(nationalId: Long,
                                        removalDate: Option[Date],
                                        inventoryDate: Date)
 
+case class TierekisteriAssetData(roadNumber: Long,
+                                 roadPartNumber: Long,
+                                 starMValue: Double,
+                                 endMValue: Double,
+                                 kvl: Int)
+
 case class TierekisteriError(content: Map[String, Any], url: String)
 
 class TierekisteriClientException(response: String) extends RuntimeException(response)
 
 class TierekisteriClientWarnings(response: String) extends RuntimeException(response)
 
-/**
-  * TierekisteriClient is a utility for using Tierekisteri (TR) bus stop REST API in OTH.
-  *
-  * @param tierekisteriRestApiEndPoint
-  * @param tierekisteriEnabled
-  *
-  */
-class TierekisteriClient(tierekisteriRestApiEndPoint: String, tierekisteriEnabled: Boolean, client: CloseableHttpClient) {
+trait TierekisteriClient{
+
+  def tierekisteriRestApiEndPoint: String
+  def tierekisteriEnabled: Boolean
+  def client: CloseableHttpClient
+
+  type TierekisteriType
 
   protected implicit val jsonFormats: Formats = DefaultFormats
+  protected val dateFormat = "yyyy-MM-dd"
+  protected val auth = new TierekisteriAuthPropertyReader
+  protected lazy val logger = LoggerFactory.getLogger(getClass)
+
+  def mapFields(data: Map[String, Any]): TierekisteriType
+
+  protected def request[T](url: String): Either[T, TierekisteriError] = {
+    val request = new HttpGet(url)
+    request.addHeader("X-OTH-Authorization", "Basic " + auth.getAuthInBase64)
+    val response = client.execute(request)
+    try {
+      val statusCode = response.getStatusLine.getStatusCode
+      if (statusCode == HttpStatus.SC_NOT_FOUND) {
+        return Right(null)
+      } else if (statusCode >= HttpStatus.SC_BAD_REQUEST) {
+        return Right(TierekisteriError(Map("error" -> ErrorMessageConverter.convertJSONToError(response), "content" -> response.getEntity.getContent), url))
+      }
+      Left(parse(StreamInput(response.getEntity.getContent)).values.asInstanceOf[T])
+    } catch {
+      case e: Exception => Right(TierekisteriError(Map("error" -> e.getMessage, "content" -> response.getEntity.getContent), url))
+    } finally {
+      response.close()
+    }
+  }
+
+  protected def post(url: String, trMassTransitStop: TierekisteriType, createJson: (TierekisteriType) => StringEntity): Option[TierekisteriError] = {

```

Comment:
Should trMassTransitStop be named something else, because it's not just for bus stops here?